### PR TITLE
RandomImprove : Test covering fees correctly

### PR DESCRIPTION
**Pull request by vsubhuman**
Added test that consistently fails on random improve not adding enough funds to cover the fees

### CODE REVIEW

**Review by vsubhuman — COMMENTED** [review]
/check

**Comment by vsubhuman** [comment]
Fails with insufficient funds at the moment:

image

Waiting for https://github.com/Emurgo/cardano-serialization-lib/issues/252 to get resolved to fix that.

The conditions in the test are:
1. Output value is 100
2. Min fee when only output is added is 53
3. Three outputs are available, 150 each
4. When only two outputs are added (which is enough to cover the output) - the min fee is 228
5. The total then comes to 328 and two outputs are not enough to cover that - **hence the fail**
6. If three outputs would be added - min fee then is 264 and the total comes to 364
7. Three outputs together are 450 and should be enough to cover the required 364

**Comment by pedromtcosta** [comment]
@vsubhuman I made the following changes:
- updated test to assert for the fee amount of `264` because now we expect the TX from that test to have 3 inputs;
- after "Phase 2: Improvement", if the total input is smaller than the total output, meaning the added fees from the inputs were not covered, we add the sufficient inputs to cover for the total output

File: rust/src/tx_builder.rs [hunk 1]
```diff
@@ -240,6 +256,23 @@ impl TransactionBuilder {
                         }
                     }
                 }
+                // Phase 3: add extra inputs needed for fees
+                // We do this at the end because this new inputs won't be associated with
+                // a specific output, so the improvement algorithm we do above does not apply here.
+                if input_total.partial_cmp(&output_total).unwrap().eq(&Ordering::Less) {
```
Comment (vsubhuman) [inline rust/src/tx_builder.rs:262]: 1. Why using `partial_cmp` here?
2. Unchecked unwrap doesn't look good.

If these two types are both `Value` I think you can even just use `<`, e.g. see how `while added < needed` is used above

Comment (pedromtcosta) [reply rust/src/tx_builder.rs]: Ah, that was a bad copy paste from my end 😅
Just pushed a version with a regular `>` comparison

**Comment by vsubhuman** [comment]
@rooooooooob , check solution plz if you have a chance

**Review by rooooooooob — APPROVED** [review]
I guess this is a good enough fix to do it at the end. CIP 2 actually leaves out the fees entirely and says it's out of scope of the CIP so this detail is not covered which is why I tried to do it during the adding phase.

We're also trying to figure out what we want to do for multiassets since those also weren't covered but we can cover that at a later point.

**Review by vsubhuman — COMMENTED** [review]
/check

**Comment by vsubhuman** [comment]
Thank you, @pedromtcosta ! Look good!

**Pull request merged**